versioned/storage/store/src/main/java/org/projectnessie/versioned/storage/versionstore/BaseMergeTransplantSquash.java

```diff
@@ -164,7 +162,11 @@ private CreateCommit createSquashCommit(
     CommitLogic commitLogic = commitLogic(persist);
     PagedResult<DiffEntry, StoreKey> diff =
         commitLogic.diff(
-            diffQuery(sourceCommits.sourceParent, sourceCommits.mostRecent(), true, filter));
+            diffQuery(
+                mergeTransplantContext.baseCommit(),
+                mergeTransplantContext.headCommit(),
```

Comment:
I tend to think a preferable approach would be to do the diff on the commit histories (similar to `git log main..dev`) as opposed to specific index states at the end of branches. Then we could aggregate incremental changes from commits not yet included into the target branch and use them to make the merge commit.

Reply:
Collecting and applying "all the commits to be merged" is extremely difficult, and also not the expected/wanted behavior. That will eventually end up in the diff anyway, because the diff is already the result of "applying all the commits".
It is not as easy as `git log main..dev` or the like - it has to traverse _all_ possible graphs of all the merges that happened in between, which is way more complicated and not necessary.

Reply:
>  I think conflict should not occur in this case, because B has already been merged into the history of G. WDYT?

That's a different issue - it*s applying the exact same thing again - that should not cause a conflict but just ignore that change.

Reply:
I guess, you're right :+1: